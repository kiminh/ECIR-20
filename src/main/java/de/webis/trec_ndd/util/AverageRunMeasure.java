package de.webis.trec_ndd.util;

import java.util.List;
import java.util.Map;

import org.apache.commons.lang3.tuple.Pair;

import com.google.common.collect.ImmutableMap;

import lombok.Data;

@Data
public class AverageRunMeasure {
	private final List<Pair<String, Double>> originalRanking;

	public Map<String, Double> reportAvg(List<Pair<String, Double>> ranking) {
		double origAvg = avg(originalRanking);
		double newAvg = avg(ranking);
		
		return ImmutableMap.<String, Double>builder()
			.put("avgOriginal", origAvg)
			.put("avgManipulated", newAvg)
			.put("difference", origAvg - newAvg)
			.build();
	}

	private static final double avg(List<Pair<String, Double>> ranking) {
		return ranking.stream()
			.mapToDouble(Pair::getRight)
			.average().getAsDouble();
	}
}
