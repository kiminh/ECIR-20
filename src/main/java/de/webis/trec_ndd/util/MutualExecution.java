package de.webis.trec_ndd.util;

import de.webis.trec_ndd.spark.QrelConsistentMaker;
import de.webis.trec_ndd.spark.RunResultDeduplicator;

public class MutualExecution {

	public static Boolean shouldExecuteTogether(QrelConsistentMaker qrel, RunResultDeduplicator dedup) {
		return !"maxValueAllDuplicateDocsIrrelevant".equals(qrel.name()) || !"base".equals(dedup.name());
	}

}
