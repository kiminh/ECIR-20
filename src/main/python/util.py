def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

