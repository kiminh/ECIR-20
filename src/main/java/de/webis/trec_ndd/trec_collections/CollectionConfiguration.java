package de.webis.trec_ndd.trec_collections;

import java.io.InputStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;

public interface CollectionConfiguration extends Serializable {

	public String getPathToCollection();
	public String getCollectionType();
	public String getDocumentGenerator();
	public List<SharedTask> getSharedTasks();

	@SneakyThrows
	public default Set<String> judgedDocumentIds() {
		Set<String> ret = new HashSet<>();
		
		for(String qrelResource : getQrelResourcesOfAllSharedTasks()) {
			InputStream resource = AnseriniCollectionReader.class.getResourceAsStream(qrelResource);
			ret.addAll(new QrelReader().readJudgedDocumentIdsFromQrels(resource));
		}
		
		return ret;
	}
	
	public default HashSet<String> documentIdsInRunFiles() {
		HashSet<String> ret = new HashSet<>();
		
		for(SharedTask sharedTask: getSharedTasks()) {
			ret.addAll(sharedTask.documentIdsInRunFiles());
		}
		
		return ret;
	}
	
	public default List<String> getQrelResourcesOfAllSharedTasks() {
		return getSharedTasks().stream()
				.map(SharedTask::getQrelResource)
				.collect(Collectors.toList());
	}
	
	@Getter
	@AllArgsConstructor
	public static enum TrecCollections implements CollectionConfiguration {
		CLUEWEB09(
			// compressed 4TB
			"/mnt/ceph/storage/corpora/corpora-thirdparty/corpus-clueweb09/",
			"ClueWeb09Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
				TrecSharedTask.WEB_2009,
				TrecSharedTask.WEB_2010,
				TrecSharedTask.WEB_2011,
				TrecSharedTask.WEB_2012,
				TrecSharedTask.SESSION_2010,
				TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012
			))
		),
		CLUEWEB12(
			"/mnt/ceph/storage/data-in-progress/kibi9872/clueweb12/",
			"ClueWeb12Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.WEB_2013,
					TrecSharedTask.WEB_2014,
					TrecSharedTask.SESSION_2013,
					TrecSharedTask.SESSION_2014
			))
		),
		GOV1(
			// compressed 4.6 GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Collections.emptyList())
		),
		GOV2(
			// compressed 81GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.TERABYTE_2004,
					TrecSharedTask.TERABYTE_2005_ADHOC,
					TrecSharedTask.TERABYTE_2006
			))
		),
		GOV2_MQ(			// compressed 81GB
				"/mnt/ceph/storage/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
				"TrecwebCollection",
				"JsoupGenerator",
				Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.MILLION_QUERY_2007,
					TrecSharedTask.MILLION_QUERY_2008
				))
		),
		ROBUST04(
			"/mnt/ceph/storage/data-in-progress/kibi9872/robust-04/",
			"TrecCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.ROBUST_04))
		),
		CORE2018(
			"/mnt/nfs/webis20/data-in-progress/trec-2018-webis/WashingtonPost/WashingtonPost.v2/data/",
			"WashingtonPostCollection",
			"WapoGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2018))
		),
		CORE2017(
			"/mnt/nfs/webis20/data-in-progress/trec-2017-webis/data/nyt/",
			"NewYorkTimesCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2017))
		);
		

		private final String pathToCollection,
							collectionType,
							documentGenerator;

		private final List<SharedTask> sharedTasks;
	}
	
	@Getter
	@AllArgsConstructor
	public static enum OtherCollections implements CollectionConfiguration {
		COMMON_CRAWL_2015_02("CommonCrawlCollection", "JsoupGenerator");
		
		private final String collectionType,
			documentGenerator;

		@Override
		public List<SharedTask> getSharedTasks() {
			return Collections.emptyList();
		}

		@Override
		public String getPathToCollection() {
			return null;
		}
	}
}
