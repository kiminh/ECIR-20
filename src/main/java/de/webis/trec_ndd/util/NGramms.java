package de.webis.trec_ndd.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Pattern;

import de.webis.trec_ndd.similarity.MD5;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

public class NGramms {

	private static final Pattern WHITESPACE = Pattern.compile("\\s+");
	
	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class Word8Gramm implements Serializable {
		private String nGramm;
		private String md5Hash;
	}
	
	public static List<Word8Gramm> build8Gramms(String text) {
		return build8Gramms(tokenize(text));
	}
	
	public static List<String> tokenize(String text) {
		return Arrays.asList(WHITESPACE.split(text));
	}
	
	public static List<Word8Gramm> build8Gramms(List<String> tokens) {
		int nGrammLength = 8;
		List<Word8Gramm> ret = new LinkedList<>();
		
		for(int i=0; i<= tokens.size() - nGrammLength; i++) {
			ret.add(buildWord8GramStartingAt(i, tokens));
		}
		
		return ret;
	}
	
	private static Word8Gramm buildWord8GramStartingAt(int position, List<String> tokens) {
		String ret = "";
		
		for(int i = 0; i< 8; i++) {
			ret += tokens.get(position + i) + " ";
		}
		
		ret = ret.trim();
		
		return new Word8Gramm(ret, MD5.md5hash(ret));
	}
}
