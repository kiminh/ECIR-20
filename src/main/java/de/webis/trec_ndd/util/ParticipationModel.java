package de.webis.trec_ndd.util;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import org.apache.commons.lang3.tuple.Pair;

import static de.webis.trec_ndd.util.ScoringChanges.failIfScoresContainsDuplicatedRuns;
import static de.webis.trec_ndd.util.ScoringChanges.failIfUnknownElementsAreInRuns;
import static de.webis.trec_ndd.util.ScoringChanges.sortedStream;

public class ParticipationModel {
	private final List<Pair<String, Double>> originalScores;
	
	public ParticipationModel(List<Pair<String, Double>> originalScores) {
		failIfScoresContainsDuplicatedRuns(originalScores);
		this.originalScores = originalScores;
	}
	public List<Integer> calculateCheatingVector(List<Pair<String, Double>> manipulatedScores) {
		failIfScoresContainsDuplicatedRuns(manipulatedScores);
		failIfUnknownElementsAreInRuns(originalScores, manipulatedScores);
		
		return sortedStream(manipulatedScores)
				.map(i -> calculateCheatingScoreForEntry(i, manipulatedScores))
				.collect(Collectors.toList());
	}
	
	public List<Integer> calculateIdealVector(List<Pair<String, Double>> manipulatedScores) {
		failIfScoresContainsDuplicatedRuns(manipulatedScores);
		failIfUnknownElementsAreInRuns(originalScores, manipulatedScores);
		
		return sortedStream(originalScores)
				.map(i -> calculateIdealScoreForEntry(extractRunWithName(i.getKey(), manipulatedScores), originalScores))
				.collect(Collectors.toList());
	}

	private Integer calculateCheatingScoreForEntry(Pair<String, Double> entry, List<Pair<String, Double>> manipulatedScores) {
		String cheater = entry.getKey();
		
		Map<String, Integer> baselinePositions = runToPosition(manipulatedScores);
		List<Pair<String, Double>> runsWithCheater = new LinkedList<>(manipulatedScores).stream()
				.filter(i -> !cheater.equals(i.getKey()))
				.collect(Collectors.toList());
		runsWithCheater.add(cheaterScore(cheater));
		
		Map<String, Integer> positionsWithSingleCheater = runToPosition(runsWithCheater);
		
		return baselinePositions.get(cheater) - positionsWithSingleCheater.get(cheater);
	}
	
	private Integer calculateIdealScoreForEntry(Pair<String, Double> entry, List<Pair<String, Double>> originalScores) {
		String ideal = entry.getKey();
		Map<String, Integer> baselinePositions = runToPosition(originalScores);
		List<Pair<String, Double>> runsWithIdeal = new LinkedList<>(originalScores).stream()
				.filter(i -> !ideal.equals(i.getKey()))
				.collect(Collectors.toList());
		runsWithIdeal.add(entry);

		Map<String, Integer> positionsWithSingleIdeal = runToPosition(runsWithIdeal);
		
		return baselinePositions.get(ideal) - positionsWithSingleIdeal.get(ideal); 
	}
	
	private Pair<String, Double> cheaterScore(String cheater) {
		return extractRunWithName(cheater, originalScores);
	}
	
	private static Pair<String, Double> extractRunWithName(String runName, List<Pair<String, Double>> runs) {
		List<Pair<String, Double>> ret = runs.stream()
				.filter(i -> runName.equals(i.getKey()))
				.collect(Collectors.toList());
		
		if(ret.size() != 1) {
			throw new RuntimeException("");
		}
		
		return ret.get(0);
	}
	
	private static Map<String, Integer> runToPosition(List<Pair<String, Double>> runs) {
		failIfScoresContainsDuplicatedRuns(runs);
		List<String> ranking = sortedStream(runs)
				.map(Pair::getLeft)
				.collect(Collectors.toList());
		
		return IntStream.range(0, ranking.size())
				.mapToObj(i -> Pair.of(ranking.get(i), i))
				.collect(Collectors.toMap(Pair::getKey, Pair::getValue));
	}
}
