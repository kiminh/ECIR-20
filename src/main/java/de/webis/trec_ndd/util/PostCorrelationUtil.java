package de.webis.trec_ndd.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiFunction;

import org.apache.commons.math3.stat.correlation.KendallsCorrelation;
import org.apache.commons.math3.stat.correlation.SpearmansCorrelation;
import org.apache.commons.math3.util.Precision;

import java.util.stream.Collectors;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import lombok.experimental.UtilityClass;

@UtilityClass
public class PostCorrelationUtil
{
	@Data
	@NoArgsConstructor
	@Accessors(chain=true)
	public static class Post {
		String id;
	}
	
	public static void main(String[] args) {
		List<Post> a = Arrays.asList(
			new Post().setId("f"),
			new Post().setId("s"),
			new Post().setId("v"),
			new Post().setId("r"),
			new Post().setId("b")
		);
		
		List<Post> b = Arrays.asList(
			new Post().setId("f"),
			new Post().setId("s"),
			new Post().setId("v"),
			new Post().setId("r"),
			new Post().setId("a")
		);
		
		System.out.println(unionSpearmanCorrelation(a,b));
	}
	
	public static double unionKendallTauCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<Post> expandedFirstRanking = expandFirstListWithEntriesFromSecond(firstRanking, secondRanking);
		List<Post> expandedSecondRanking = expandFirstListWithEntriesFromSecond(secondRanking, firstRanking);
		
		return intersectionKendallTauCorrelation(expandedFirstRanking, expandedSecondRanking);
	}
	
	public static double intersectionKendallTauCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		return calculateCorrelation(firstRanking, secondRanking, new KendallsCorrelation()::correlation);
	}
	
	public static double unionSpearmanCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<Post> expandedFirstRanking = expandFirstListWithEntriesFromSecond(firstRanking, secondRanking);
		List<Post> expandedSecondRanking = expandFirstListWithEntriesFromSecond(secondRanking, firstRanking);
		
		return intersectionSpearmanCorrelation(expandedFirstRanking, expandedSecondRanking);
	}
	
	public static double intersectionSpearmanCorrelation(List<Post> firstRanking, List<Post> secondRanking)
	{
		return calculateCorrelation(firstRanking, secondRanking, new SpearmansCorrelation()::correlation);
	}
	
	private static double calculateCorrelation(List<Post> firstRanking, List<Post> secondRanking, BiFunction<double[], double[], Double> correlation)
	{
		double[] firstRankingArray = buildCorrelationArrayForFirstRanking(firstRanking, secondRanking);
		double[] secondRankingArray = buildCorrelationArrayForFirstRanking(secondRanking, firstRanking);
		
		if(firstRankingArray.length == 0)
		{
			return 0;
		}
		else if(firstRankingArray.length == 1)
		{
			return Precision.equals(firstRankingArray[0], secondRankingArray[0]) ? 1 : 0;
		}
		
		return correlation.apply(firstRankingArray, secondRankingArray);
	}
	
	private static double[] buildCorrelationArrayForFirstRanking(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<String> firstRankingIds = mapToIds(firstRanking);
		firstRankingIds.retainAll(mapToIds(secondRanking));
		
		List<String> indexValues = new ArrayList<>(firstRankingIds);
		indexValues.sort(String::compareTo);
		
		return indexValues.stream()
				.map(id -> firstRankingIds.indexOf(id))
				.mapToDouble(Integer::doubleValue)
				.toArray();
	}
	
	private static List<Post> expandFirstListWithEntriesFromSecond(List<Post> firstRanking, List<Post> secondRanking)
	{
		List<String> firstRankingIds = mapToIds(firstRanking);
		List<String> secondRankingIds = mapToIds(secondRanking);
		
		secondRankingIds.removeAll(firstRankingIds);
		
		if(secondRankingIds == null || secondRanking == null || secondRankingIds.size() == secondRanking.size())
		{
			return new ArrayList<>();
		}
		
		firstRankingIds.addAll(secondRankingIds);
		
		return firstRankingIds.stream()
				.map(id -> new Post().setId(id))
				.collect(Collectors.toList());
	}
	
	private static List<String> mapToIds(List<Post> posts)
	{
		if(posts == null)
		{
			return new ArrayList<>();
		}
		
		List<String> ret = posts.stream().map(Post::getId).collect(Collectors.toList());
		
		if(ret.stream().anyMatch(a -> a == null || a.isEmpty()) || ret.stream().distinct().collect(Collectors.toList()).size() != ret.size())
		{
			throw new IllegalArgumentException("The following list of posts contains duplicates or missing ids. "
					+ "I dont know how to calculate a correlation coefficient: "+ ret);
		}
		
		return ret;
	}
}
