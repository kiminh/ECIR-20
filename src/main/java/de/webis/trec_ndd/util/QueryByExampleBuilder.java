package de.webis.trec_ndd.util;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.standard.StandardAnalyzer;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class QueryByExampleBuilder {
	public static String esQueryByExample(String document){
		return esOrQuery(tokensInText(document));
	}
	
	public static String esQueryByExampleWithKMedianTokens(String document, int k){
		List<String> sortedTokens = new ArrayList<String>(tokensInText(document));
		Collections.sort(sortedTokens, (a,b) -> {
			try {
				return WordCounts.getWordCount(a).compareTo(WordCounts.getWordCount(b));
			} catch (ExecutionException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
				System.exit(1);
				return 1;
			}
		});
		Set<String> ret = new HashSet<>();
		
		while(k > ret.size() && sortedTokens.size() > 0) {
			ret.add(sortedTokens.remove(sortedTokens.size()/2));
		}
		
		return esOrQuery(ret);
	}
	
	private static String esOrQuery(Collection<String> tokens) {
		return tokens.stream()
			.map(t -> "(body_lang.en: \"" + t + "\")")
			.collect(Collectors.joining(" OR "));
	}

	private static Set<String> tokensInText(String text) {
		List<String> ret = textToTokenList(text);
		return new HashSet<>(ret);
	}

	private static List<String> tokensInText(Analyzer analyzer, String text) throws IOException {
		List<String> ret = new LinkedList<>();
		TokenStream tokenStream = analyzer.tokenStream("", text);

		CharTermAttribute attr = tokenStream.addAttribute(CharTermAttribute.class);
		tokenStream.reset();

		while (tokenStream.incrementToken()) {
			ret.add(attr.toString());
		}

		return ret;
	}

	@SneakyThrows
	public static List<String> textToTokenList(String text) {
		Analyzer analyzer = new StandardAnalyzer();
		
		return tokensInText(analyzer, text);
	}
}
