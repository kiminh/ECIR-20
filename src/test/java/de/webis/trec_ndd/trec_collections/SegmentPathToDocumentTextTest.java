package de.webis.trec_ndd.trec_collections;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.approvaltests.Approvals;
import org.junit.Test;
import org.mockito.Mockito;

import com.google.common.collect.Lists;

import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;

public class SegmentPathToDocumentTextTest {

	@Test
	public void approveTransformationForRobustCollectionWithSingleSegment() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/robust", TrecCollections.ROBUST_TBD);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForClueweb09CollectionWithSingleSegment() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/clueweb09-sample", TrecCollections.CLUEWEB09);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForClueweb09CollectionWithMultipleSegments() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/clueweb09-sample-2", TrecCollections.CLUEWEB09);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForGOV1CollectionWithMultipleSegments() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/gov1-sample", TrecCollections.GOV1);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForGOV2CollectionWithMultipleSegments() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/gov2-sample", TrecCollections.GOV2);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForCore2018CollectionWithSingleSegments() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/core-2018-sample", TrecCollections.CORE2018);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForCore2017CollectionWithMultipleSegments() {
		CollectionConfiguration conf = collectionConfiguration("src/test/resources/data/core-2017-sample", TrecCollections.CORE2017);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<CollectionDocument> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		Collections.sort(collectionDocuments, (a,b) -> a.getId().compareTo(b.getId()));
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	private static List<CollectionDocument> mapAllSegmentPathsToDocumentRepresentation(AnseriniCollectionReader<?> collectionReader) {
		return collectionReader.segmentPaths().stream()
				.map(i -> collectionReader.collectionDocumentsInPath(i))
				.map(Lists::newArrayList)
				.flatMap(List::stream)
				.collect(Collectors.toList());
	}
	
	public static CollectionConfiguration collectionConfiguration(String pathToCollection, TrecCollections collection) {
		return collectionConfiguration(pathToCollection, Collections.emptyList(), collection);
	}
	
	public static CollectionConfiguration collectionConfiguration(String pathToCollection, List<String> qrelResources, TrecCollections collection ) {
		List<SharedTask> sharedTasks = qrelResources.stream()
				.map(r -> sharedTaskWithQrelResource(r))
				.collect(Collectors.toList());
		
		return collectionConfigurationWithSharedTasks(pathToCollection, sharedTasks, collection);
	}

	@SuppressWarnings("serial")
	public static CollectionConfiguration collectionConfigurationWithSharedTasks(String pathToCollection, List<SharedTask> sharedTasks, TrecCollections collection ) {
		return new CollectionConfiguration() {
			@Override
			public String getPathToCollection() {
				return pathToCollection;
			}

			@Override
			public String getCollectionType() {
				return collection.getCollectionType();
			}

			@Override
			public String getDocumentGenerator() {
				return collection.getDocumentGenerator();
			}

			@Override
			public List<SharedTask> getSharedTasks() {
				return sharedTasks;
			}
		};
	}

	
	private static SharedTask sharedTaskWithQrelResource(String qrelResource) {
		SharedTask ret = Mockito.mock(SharedTask.class);
		Mockito.when(ret.getQrelResource()).thenReturn(qrelResource);
		
		return ret;
	}
}
