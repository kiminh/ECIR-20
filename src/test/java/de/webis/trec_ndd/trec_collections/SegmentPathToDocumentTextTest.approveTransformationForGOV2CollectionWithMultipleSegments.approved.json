[
  {
    "id": "GX000-00-0000000",
    "content": "JPL Sgra web Site\n JPL Sgra Web Site The US Space VLBI project web site has moved. Click HERE to go to the new US Space VLBI project web site. The following web pages are now available from this site: Old US space VLBI project web site Project science web page Last updated: Thu Sep 16 17:24:48 PDT 1999",
    "fullyCanonicalizedContent": "jpl sgra web site jpl sgra web site u space vlbi project web site ha move click here go new u space vlbi project web site follow web page now avail from site old u space vlbi project web site project scienc web page last updat thu sep 16 17 24 48 pdt 1999"
  }
]